Modules/_copy.c

```diff
@@ -0,0 +1,432 @@
+#ifndef Py_BUILD_CORE_BUILTIN
+#  define Py_BUILD_CORE_MODULE 1
+#endif
+#define NEEDS_PY_IDENTIFIER
+
+#include "Python.h"
+
+/*[clinic input]
+module _copy
+[clinic start generated code]*/
+/*[clinic end generated code: output=da39a3ee5e6b4b0d input=b34c1b75f49dbfff]*/
+
+/*
+ * Duplicate of builtin_id. Looking around the CPython code base, it seems to
+ * be quite common to just open-code this as PyLong_FromVoidPtr, though I'm
+ * not sure those cases actually need to interoperate with Python code that
+ * uses id(). We do, however, so it would be nicer there was an official
+ * public API (e.g. PyObject_Id, maybe just a macro to avoid extra
+ * indirection) providing this..
+ */
+static PyObject*
+object_id(PyObject* v)
+{
+    return PyLong_FromVoidPtr(v);
+}
+
+static int
+memo_keepalive(PyObject* x, PyObject* memo)
+{
+    PyObject* memoid, * list;
+    int ret;
+
+    memoid = object_id(memo);
+    if (memoid == NULL)
+        return -1;
+
+    /* try: memo[id(memo)].append(x) */
+    list = PyDict_GetItem(memo, memoid);
+    if (list != NULL) {
+        Py_DECREF(memoid);
+        return PyList_Append(list, x);
+    }
+
+    /* except KeyError: memo[id(memo)] = [x] */
+    list = PyList_New(1);
+    if (list == NULL) {
+        Py_DECREF(memoid);
+        return -1;
+    }
+    Py_INCREF(x);
+    PyList_SET_ITEM(list, 0, x);
+    ret = PyDict_SetItem(memo, memoid, list);
+    Py_DECREF(memoid);
+    Py_DECREF(list);
+    return ret;
+}
+
+/* Forward declaration. */
+static PyObject* do_deepcopy(PyObject* x, PyObject* memo);
+
+static PyObject*
+do_deepcopy_fallback(PyObject* x, PyObject* memo)
+{
+    static PyObject* copymodule;
```

Comment:
You must not cache Python objects in local static variables. It is incompatible with subinterpreters and causes memory leaks on shutdown. If you want to cache the copy module then you have to use module state API.